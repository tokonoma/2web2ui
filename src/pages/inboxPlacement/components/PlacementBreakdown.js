import React from 'react';
import { PanelSectionTableCollection } from 'src/components/collection';
import { PageLink } from 'src/components/links';
import { slugToFriendly } from 'src/helpers/string';
import { formatPercent } from 'src/helpers/units';
import { PLACEMENT_FILTER_TYPES } from '../constants/types';
import formatFilterName, { formatRegion } from '../helpers/formatFilterName.js';

const getPlacementNameByType = (type, { mailbox_provider, region, sending_ip }) => {
  switch (type) {
    case PLACEMENT_FILTER_TYPES.REGION:
      return region;
    case PLACEMENT_FILTER_TYPES.SENDING_IP:
      return sending_ip;
    case PLACEMENT_FILTER_TYPES.MAILBOX_PROVIDER:
      return mailbox_provider;
    default:
      return '';
  }
};

const PlacementBreakdown = ({ data = [], type }) => {
  const columns = [
    {
      component: props => {
        const name = getPlacementNameByType(type, props);

        return (
          <PageLink to={`/inbox-placement/details/${props.id}/${type}/${name}`}>
            {formatFilterName(type, name)}
          </PageLink>
        );
      },
      header: {
        label: slugToFriendly(type),
        sortKey: props => (getPlacementNameByType(type, props) || '').toLowerCase(),
      },
    },
    {
      component: ({ region }) => formatRegion(region),
      header: {
        label: 'Region',
        sortKey: 'region',
      },
      visible: () => type === PLACEMENT_FILTER_TYPES.MAILBOX_PROVIDER,
    },
    {
      component: ({ placement }) => formatPercent(placement.inbox_pct * 100),
      header: {
        label: 'Inbox',
        sortKey: 'placement.inbox_pct',
        minWidth: '72px',
      },
    },
    {
      component: ({ placement }) => formatPercent(placement.spam_pct * 100),
      header: {
        label: 'Spam',
        sortKey: 'placement.spam_pct',
        minWidth: '72px',
      },
    },
    {
      component: ({ placement }) => formatPercent(placement.missing_pct * 100),
      header: {
        label: 'Missing',
        sortKey: 'placement.missing_pct',
        minWidth: '72px',
      },
    },
  ];
  const visibleColumns = columns.filter(({ visible = () => true }) => visible());
  const renderRow = columns => props =>
    columns.map(({ component: Component }) => <Component {...props} />);

  return (
    <PanelSectionTableCollection
      columns={visibleColumns.map(({ header }) => header)}
      rows={data}
      getRowData={renderRow(visibleColumns)}
      defaultSortColumn="placement.inbox_pct"
      defaultSortDirection="desc"
      filterBox={false}
      pagination={true}
      saveCsv={true}
    />
  );
};

export default PlacementBreakdown;
