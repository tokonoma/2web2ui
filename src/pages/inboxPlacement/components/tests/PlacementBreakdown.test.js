import React from 'react';
import { shallow } from 'enzyme';
import PlacementBreakdown from '../PlacementBreakdown';

describe('Component: PlacementBreakdown', () => {
  const data = [
    {
      mailbox_provider: 'dogmail.com',
      region: 'north america',
      sending_ip: '101.101',
      placement: {
        inbox_pct: 0,
        spam_pct: 0,
        missing_pct: 1,
      },
      authentication: {
        spf_pct: 0,
        dkim_pct: 0,
        dmarc_pct: 1,
      },
    },
    {
      mailbox_provider: 'doghouseinbox.com',
      region: 'europe',
      sending_ip: '101.102',
      placement: {
        inbox_pct: 0.8,
        spam_pct: 0.1,
        missing_pct: 0.1,
      },
      authentication: {
        spf_pct: 0.8,
        dkim_pct: 0.1,
        dmarc_pct: 0.1,
      },
    },
  ];

  const subject = ({ ...props }) => shallow(<PlacementBreakdown data={[]} {...props} />);

  it('renders correctly with no data', () => {
    expect(subject()).toMatchSnapshot();
  });

  it('matches snapshot with provider data', () => {
    expect(subject({ data })).toMatchSnapshot();
  });
});
