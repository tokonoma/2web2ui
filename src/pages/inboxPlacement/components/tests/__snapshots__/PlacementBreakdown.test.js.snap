// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: PlacementBreakdown matches snapshot with provider data 1`] = `
<PanelSectionTableCollection
  columns={
    Array [
      Object {
        "label": "",
        "sortKey": [Function],
      },
      Object {
        "label": "Inbox",
        "minWidth": "72px",
        "sortKey": "placement.inbox_pct",
      },
      Object {
        "label": "Spam",
        "minWidth": "72px",
        "sortKey": "placement.spam_pct",
      },
      Object {
        "label": "Missing",
        "minWidth": "72px",
        "sortKey": "placement.missing_pct",
      },
    ]
  }
  defaultSortColumn="placement.inbox_pct"
  defaultSortDirection="desc"
  filterBox={false}
  getRowData={[Function]}
  pagination={true}
  rows={
    Array [
      Object {
        "authentication": Object {
          "dkim_pct": 0,
          "dmarc_pct": 1,
          "spf_pct": 0,
        },
        "mailbox_provider": "dogmail.com",
        "placement": Object {
          "inbox_pct": 0,
          "missing_pct": 1,
          "spam_pct": 0,
        },
        "region": "north america",
        "sending_ip": "101.101",
      },
      Object {
        "authentication": Object {
          "dkim_pct": 0.1,
          "dmarc_pct": 0.1,
          "spf_pct": 0.8,
        },
        "mailbox_provider": "doghouseinbox.com",
        "placement": Object {
          "inbox_pct": 0.8,
          "missing_pct": 0.1,
          "spam_pct": 0.1,
        },
        "region": "europe",
        "sending_ip": "101.102",
      },
    ]
  }
  saveCsv={true}
/>
`;

exports[`Component: PlacementBreakdown renders correctly with no data 1`] = `
<PanelSectionTableCollection
  columns={
    Array [
      Object {
        "label": "",
        "sortKey": [Function],
      },
      Object {
        "label": "Inbox",
        "minWidth": "72px",
        "sortKey": "placement.inbox_pct",
      },
      Object {
        "label": "Spam",
        "minWidth": "72px",
        "sortKey": "placement.spam_pct",
      },
      Object {
        "label": "Missing",
        "minWidth": "72px",
        "sortKey": "placement.missing_pct",
      },
    ]
  }
  defaultSortColumn="placement.inbox_pct"
  defaultSortDirection="desc"
  filterBox={false}
  getRowData={[Function]}
  pagination={true}
  rows={Array []}
  saveCsv={true}
/>
`;
